bftengine/tests/testSeqNumForClientRequest/seqNumForClientRequest_test.cpp

```diff
@@ -1,8 +1,73 @@
 #include "gtest/gtest.h"
+#include "SimpleClientImp.cpp"
 
-TEST(seqNumForClientRequest_test, initial_test) {}
+using namespace std;
+using namespace bftEngine;
 
-int main(int argc, char* argv[]) {
+namespace {
+
+class MockSeqNumGenerator : public SeqNumberGeneratorForClientRequestsImp {
+ public:
+  uint32_t getLastCountOfUniqueFetchID() { return lastCountOfUniqueFetchID_; }
+
+  void setLastCountOfUniqueFetchID(uint64_t count) { lastCountOfUniqueFetchID_ = count; }
+};
+
+TEST(seqNumForClientRequest_test, send_requests_same_time_uniqueness_preserved) {
+  SeqNumberGeneratorForClientRequests *seqNumGen = new MockSeqNumGenerator();
+  std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
+  set<u_int64_t> seqNums;
+
+  for (int i = 0; i < 10; i++) {
+    seqNums.insert(seqNumGen->generateUniqueSequenceNumberForRequest(now));
+  }
+
+  // assert 10 unique SN was generated, although practically sent at the same time.
+  Assert(seqNums.size() == 10);
+  delete seqNumGen;
+}
+
+TEST(seqNumForClientRequest_test, exceed_counter_limit_uniqueness_preserved) {
+  SeqNumberGeneratorForClientRequests *seqNumGen = new MockSeqNumGenerator();
+  std::chrono::time_point<std::chrono::system_clock> now = std::chrono::system_clock::now();
+  set<u_int64_t> seqNums;
+
+  seqNums.insert(seqNumGen->generateUniqueSequenceNumberForRequest(now));
+  auto mockSeqNumGen = dynamic_cast<MockSeqNumGenerator *>(seqNumGen);
+
+  // Set lastCount biggest value.
+  mockSeqNumGen->setLastCountOfUniqueFetchID(0x3FFFFF);
```

Comment:
Where does this 0x3FFFFF number come from?

Reply:
The class generate SN out of time_point.
In order to preserve the uniqueness, we use to variables:
last_milli -> the sampled time_point.
last_count -> used to preserve uniqueness.
Before we return SN we shift to the right (0x3FFFFF), the last_milli and use bitwise 'or' with last count.
If we will try to generate another SN and the last_milli will be identical (time sampled is equal to the last one sampled), we will increase last_count by 1, and do the bitwise or with different number, therefore we will get unique SN.
The last_count is limited to increase up to 0x3FFFFF, the size we shifted the last_milli variable.

The class handle the scenario of exceeding the last_count limit, as tested in 'exceed_counter_limit_uniqueness_preserved'